[Feature] Add Websocket Feed

Sotam opened the pull request:
- Updated Newtonsoft and added Websocket4Net for the websocket feed.
- Moved `ComputeSignature` and `HashString` to `Authenticator.cs`, because the encryption method is necessary for authenticating in the user channel. Not quite sure if this is the most appropriate place?
- For all the models, the [JsonProperty] is probably not necessary?
- Used a generic EventArgs, probably it can be done better? Not really sure (line 135 - 137, WebSocket.cs)
- Sometimes the connection tends to stop, currently it'll automatically restart. Next step is implement the `CancellationToken`

Please let me know what you guys think of this :)

dougdellolio commented on GDAXSharp/Network/Authentication/Authenticator.cs:
```diff
@@ -17,5 +23,21 @@ public class Authenticator : IAuthenticator
 
         public string Passphrase { get; }
+
+        public string ComputeSignature(HttpMethod httpMethod, string secret, double timestamp, string requestUri, string contentBody = "")
```
I am of the opinion that this still belongs in the `HttpRequestMessageService` because that's really the only thing that cares about it.

Sotam replied:
I somewhat agree, however, for the authentication of the user channel this function is needed and since this function is always needed when authenticating I thought this could be an appropriate place. Also initializing `HttpRequestMessageService` just for `ComputeSignature` seems not the right way. Please let me know what you think of it :)

Sotam replied:
I somewhat agree, however, for the authentication of the user channel this function is needed and since this function is always needed when authenticating I thought this could be an appropriate place. Also initializing `HttpRequestMessageService` just for `ComputeSignature` seems not the right way. Please let me know what you think of it :)

dougdellolio commented on GDAXSharp/Shared/Types/Side.cs:
```diff
@@ -0,0 +1,7 @@
+﻿namespace GDAXSharp.Shared.Types
+{
+    public enum Side
```
this already exists here https://github.com/dougdellolio/gdax-csharp/blob/master/GDAXSharp/Services/Orders/Types/OrderSide.cs

Sotam replied:
Shouldn't this one be in the `Shared` folder? Because multiple services use this enum and not all of them are of the type `Order`

dougdellolio commented on GDAXSharp/Shared/Utilities/Extensions/ProductTypeExtensions.cs:
```diff
@@ -18,5 +18,12 @@ public static Currency QuoteCurrency(this ProductType value)
             return (Currency)Enum.Parse(typeof(Currency), quoteCurrency);
         }
+
+        public static string ToDasherizedUpper(this ProductType orderType)
```
shouldn't need this. you should be able to use the methods above here to get the currency in the format you are looking for

Sotam replied:
In your webfeed branch I saw you created a function like this. so I somewhat copied it.

dougdellolio commented on GDAXSharp/WebSocket/Models/Request/TickerChannel.cs:
```diff
@@ -0,0 +1,38 @@
+{
+    public class TickerChannel
+    {
+        [JsonProperty("type")]
```
shouldn't need all of these attributes

Sotam replied:
Isn't this necessary to build/serialize to JSON?

chrisw000 replied:
@Sotam - have a look in `AbstractService `at `protected static string SerializeObject` and the deserialize method... they both use specific `JsonSerializerSettings` which take case of the gdax naming convention automatically (in vast majority of cases)

The 2 static methods:  `SerializeObject `/ `DeserializeObject `and static property `SerializerSettings ` could be removed from  AbstractService to a static helper class... then have WebSocket and AbstractService both make use of new helper class?

dougdellolio commented on GDAXSharp/WebSocket/Models/Request/TickerChannel.cs:
```diff
@@ -0,0 +1,38 @@
+    public class TickerChannel
+    {
+        [JsonProperty("type")]
+        public string Type { get; set; }
```
you can convert these into their enums doing something like this `[JsonConverter(typeof(StringEnumConverter))]`

here is another example https://github.com/dougdellolio/gdax-csharp/blob/master/GDAXSharp/Services/Orders/Models/Responses/OrderResponse.cs

dougdellolio commented on GDAXSharp/WebSocket/Models/Request/TickerChannel.cs:
```diff
@@ -0,0 +1,38 @@
+        public string Type { get; set; }
+
+        [JsonProperty("product_ids")]
+        public List<string> ProductIds { get; set; }
```
can use the enum here similar to this:

https://github.com/dougdellolio/gdax-csharp/blob/master/GDAXSharp/Services/Orders/Models/Responses/OrderResponse.cs#L18

dougdellolio commented on GDAXSharp/WebSocket/Models/Request/TickerChannel.cs:
```diff
@@ -0,0 +1,38 @@
+        public string Passphrase { get; set; }
+
+        [JsonProperty("timestamp")]
+        public string Timestamp { get; set; }
```
should be a `DateTime`

Sotam replied:
I think not, it should be the same as this:
https://github.com/dougdellolio/gdax-csharp/blob/cc835389bc7bd531b3a833f1d398ee9616303fbd/GDAXSharp/Network/HttpRequest/HttpRequestMessageService.cs#L77 Perhaps I'm missing something?

dougdellolio replied:
oh yeah this needs to be a string

dougdellolio commented on GDAXSharp/WebSocket/Models/Response/Level2.cs:
```diff
@@ -0,0 +1,18 @@
+    public class Level2 : BaseMessage
+    {
+        [JsonProperty("product_id")]
+        public string ProductId { get; set; }
```
this can also be converted to the enum.

dougdellolio commented on GDAXSharp/WebSocket/WebSocket.cs:
```diff
@@ -0,0 +1,139 @@
+        private ProductType[] ProductTypes { get; set; }
+        private WebSocket4Net.WebSocket WebSocketFeed { get; set; }
+
+        public WebSocket(IAuthenticator authenticator, IClock clock, CancellationToken cancellationToken, bool sandBox = false)
```
i don't think the user should have to pass in an `IClock`. they shouldn't care about it imo

dougdellolio commented:
@chrisw000 mind taking a look?

Sotam commented:
Oops, I see that I've made some mistakes (duplicate types, JSON attributes). I'll make some changes later today. Thanks for your review!

chrisw000 commented:
@Sotam  Awesome!

@dougdellolio - there is some common ground between `HttpRequestMessageService `and `WebSocket` - maybe they could share a common base class `AbstractRequest` 
 -> move the `ComputeSignature `code from `HttpRequestMessageService `down into the common base (rather than having it on the `IAuthenticator`)
-> Passing in the `IClock `in follows the pattern in `HttpRequestMessageService`, so in that sense it's correct to pass into `WebSocket` - or just don't pass it into either?
-> move `IAuthenticator, IClock, sandbox` down as protected member vars of the `AbstractRequest `

if #98 gets merged before this, use the new Exceptions namespace to introduce a `GDAXSharpWebSocketException `  into `WebSocket.WebSocket_Error`

`WebSocketFeed `( in `WebSocket `) is `Disposable`, so `WebSocket ` could (should?) implement `IDisposable` ?

Sotam commented:
@dougdellolio I have made quite some (hopefully) improvements.
@chrisw000 I implemented an ~AbstractRequest~ and I hope it's what you mean. 

Please let me know your opinions.

chrisw000 commented:
@Sotam - yeah that is exactly what I meant. nice one.

`ComputeSignature` could be protected static and `HashString` could be private static

dougdellolio commented on GDAXSharp/Network/AbstractRequest.cs:
```diff
@@ -0,0 +1,40 @@
+{
+    public class AbstractRequest
+    {
+        protected readonly IAuthenticator Authenticator;
```
new line these fields

dougdellolio commented on GDAXSharp/Network/AbstractRequest.cs:
```diff
@@ -0,0 +1,40 @@
+            SandBox = sandBox;
+        }
+
+        public string ComputeSignature(HttpMethod httpMethod, string secret, double timestamp, string requestUri, string contentBody = "")
```
this can be `protected`

dougdellolio commented on GDAXSharp/Network/AbstractRequest.cs:
```diff
@@ -0,0 +1,40 @@
+            }
+        }
+    }
+}
```
add new line

dougdellolio commented on GDAXSharp/Network/HttpRequest/HttpRequestMessageService.cs:
```diff
@@ -1,39 +1,29 @@
-        private readonly bool sandBox;
-
-        public HttpRequestMessageService(IAuthenticator authenticator, IClock clock, bool sandBox)
+        public HttpRequestMessageService(IAuthenticator authenticator, IClock clock, bool sandBox) : base(authenticator, clock, sandBox)
```
please new line these

````
public HttpRequestMessageService(
    IAuthenticator authenticator, 
    IClock clock, 
    bool sandBox) 
         : base(authenticator, clock, sandBox) 
{
...
}
````

dougdellolio commented on GDAXSharp/Shared/Types/ResponseType.cs:
```diff
@@ -0,0 +1,22 @@
+        Activate,
+        Error
+    }
+}
```
new line

dougdellolio commented on GDAXSharp/WebSocket/IWebSocket.cs:
```diff
@@ -0,0 +1,15 @@
+        event EventHandler<WebfeedEventArgs<Snapshot>> OnSnapShotReceived;
+        event EventHandler<WebfeedEventArgs<Level2>> OnLevel2UpdateReceived;
+    }
+}
```
new line

dougdellolio commented on GDAXSharp/WebSocket/IWebSocket.cs:
```diff
@@ -0,0 +1,15 @@
+    {
+        void GetTickerChannel(params ProductType[] productTypes);
+
+        event EventHandler<WebfeedEventArgs<Ticker>> OnTickerReceived;
```
new line these properly

dougdellolio commented on GDAXSharp/WebSocket/Models/Response/BaseMessage.cs:
```diff
@@ -0,0 +1,9 @@
+    {
+        public ResponseType Type { get; set; }
+    }
+}
```
nl

dougdellolio commented on GDAXSharp/WebSocket/Models/Response/Level2.cs:
```diff
@@ -0,0 +1,14 @@
+
+        public List<string[]> Changes { get; set; }
+    }
+}
```
nl

dougdellolio commented on GDAXSharp/WebSocket/Models/Response/Snapshot.cs:
```diff
@@ -0,0 +1,14 @@
+
+        public List<decimal[]> Asks { get; set; }
+    }
+}
```
nl

dougdellolio commented on GDAXSharp/WebSocket/Models/Response/Ticker.cs:
```diff
@@ -0,0 +1,36 @@
+
+        public string LastSize { get; set; }
+    }
+}
```
nl

dougdellolio commented on GDAXSharp/WebSocket/Models/Response/WebSocketFeedEventArgs.cs:
```diff
@@ -0,0 +1,14 @@
+
+        public T LastOrder { get; }
+    }
+}
```
nl

dougdellolio commented on GDAXSharp/WebSocket/WebSocket.cs:
```diff
@@ -0,0 +1,135 @@
+
+        private const string SandBoxApiUri = "wss://ws-feed-public.sandbox.gdax.com";
+
+        private CancellationToken CancellationToken { get; set; }
```
please move the set of props under the constructor and new line them

dougdellolio replied:
also if you are setting this in the constructor you can remove this `set` here

dougdellolio commented on GDAXSharp/WebSocket/WebSocket.cs:
```diff
@@ -0,0 +1,135 @@
+        private ProductType[] ProductTypes { get; set; }
+        private WebSocket4Net.WebSocket WebSocketFeed { get; set; }
+
+        public WebSocket(IAuthenticator authenticator, IClock clock, CancellationToken cancellationToken, bool sandBox = false) : base(authenticator, clock, sandBox)
```
new line these guys in the constructor.

````
public WebSocket(
    IAuthenticator authenticator, 
    IClock clock, 
    CancellationToken cancellationToken,
    bool sandBox = false)
         : base(authenticator, clock, sandBox) 
{
...
}
````

dougdellolio commented on GDAXSharp/WebSocket/WebSocket.cs:
```diff
@@ -0,0 +1,135 @@
+                    new Channel
+                    {
+                        Name = "user",
+
```
remove this extra line here

dougdellolio commented on GDAXSharp/WebSocket/WebSocket.cs:
```diff
@@ -0,0 +1,135 @@
+            switch (response.Type)
+            {
+                case ResponseType.Subscriptions:
+                    // TODO: Implement this and more?
```
what is this for? if you aren't going to add this here i would remove and add in a subsequent PR

Sotam replied:
There are more response types, so I was thinking about adding different `EventHandler` for each `ResponseType`.

dougdellolio replied:
awesome. just remove the TODO if you aren't going to implement it yet.

dougdellolio commented on GDAXSharp/WebSocket/WebSocket.cs:
```diff
@@ -0,0 +1,135 @@
+        public event EventHandler<WebfeedEventArgs<Snapshot>> OnSnapShotReceived;
+        public event EventHandler<WebfeedEventArgs<Level2>> OnLevel2UpdateReceived;
+    }
+}
```
nl

dougdellolio commented on GDAXSharp/WebSocket/WebSocket.cs:
```diff
@@ -0,0 +1,135 @@
+        private ProductType[] ProductTypes { get; set; }
+        private WebSocket4Net.WebSocket WebSocketFeed { get; set; }
+
+        public WebSocket(IAuthenticator authenticator, IClock clock, CancellationToken cancellationToken, bool sandBox = false) : base(authenticator, clock, sandBox)
```
so I am still of the opinion here that the user should not have to pass in the `IClock`. to @chrisw000 's point. yes the `HttpRequestMessageService` follows the same pattern but the user doesn't care about a clock when creating a new `GDAXClient`.

in this case here. when the user wants to new up a new `Websocket` they will have to do something like this 

`var webSocket = new WebSocket(authenticator, new Clock(), CancellationToken.None);`

Sotam replied:
You are right, this needs to be changed.

chrisw000 replied:
Does the IClock only exist for the tests, to allow the test to dictate a fixed time?

If so, would another solution be to just get rid of IClock altogether, and put the GetTime() method (from clock) onto AbstractRequest - which the test could still change?

dougdellolio commented on GDAXSharp/WebSocket/WebSocket.cs:
```diff
@@ -0,0 +1,135 @@
+
+        private void WebSocket_Error(object sender, ErrorEventArgs e)
+        {
+            new Exception($"WebSocket Feed Error: {e.Exception.Message}");
```
you can use @chrisw000 code from #98. `GDAXSharpWebSocketException`

dougdellolio commented on GDAXSharp/WebSocket/Models/Response/Ticker.cs:
```diff
@@ -0,0 +1,36 @@
+
+        public long TradeId { get; set; }
+
+        public string LastSize { get; set; }
```
`decimal`?

Sotam commented:
@dougdellolio @chrisw000 I can't seem to figure why the build suddenly fails. On my machine it builds and works fine.. Travic-CI says something about missing `Newtonsoft`.

Anyway, thanks for your reviews, will make changes later today!

Sotam commented on GDAXSharp/WebSocket/WebSocket.cs:
```diff
@@ -0,0 +1,135 @@
+        private void WebSocket_Opened(object sender, EventArgs e)
+        {
+            var timeStamp = Clock.GetTime().ToTimeStamp();
+            var json = JsonConvert.SerializeObject(new TickerChannel
```
@dougdellolio @chrisw000 would you be willing to look at this? Currently it's not changeable by the users. So I was thinking, perhaps a part of could be a parameter. Users would be able to specify which channels they want to subscribe to.

chrisw000 replied:
Don't see why not. Have a flag enum to represent each channel... . allow use to pass an channels emum value into GetTickerChannel (default to All) - then use that value to determine which Channel objects to add into the List<Channel> when setting up the WebSocketFeed.

If the Channel objects property Namewere also this enum type (with correct serialization info on the enum) it would just be a loop to add each channel being passed in?

chrisw000 commented:
I pulled a copy of your branch - and it doesn't build for me, gives same error as Travis.

You've got the wrong hint paths in the `GDAXSharp.csproj file`

You currently have ../../    (up 2 folders)

It should be ../     (up 1 folder)

Once I edited the .csproj file in notepad it compiled just fine.

I guess the packages are duplicated on your machine 2 folders up so it works for you.....

Sotam commented:
@chrisw000 ahhhhh that makes sense, as I've included the repo inside another repo. It should be fixed now.

(@dougdellolio) Furthermore I've changed the feedback and added the `WebSocket` to the `GDAXClient`, I think this solves the ugly way of using the `WebSocket` functionality.  I think we are getting somewhere. 😃

dougdellolio commented on GDAXSharp/GDAXClient.cs:
```diff
@@ -15,25 +15,29 @@
             IAuthenticator authenticator,
-            bool sandBox = false) 
-                    : this(authenticator, new HttpClient(), sandBox)
+            CancellationToken cancellationToken,
```
so this is a breaking change. do we want them to always have to pass in a `CancellationToken`?

Sotam replied:
I think we do, the purpose of the `cancellationToken` is to stop the `WebSocket`. Perhaps this is not the appropriate way, perhaps a boolean parameter is better way?

chrisw000 replied:
Maybe don't pass the cancellationToken in via the WebSocket constructor.... meaning you don't need it passing into the GDAXClient.

Pass the cancellationToken in here:
 public void Start(List<ProductType> productTypes, List<ChannelType> channelTypes = null)

Or hold the cancellationToken internally within WebSocket and provide a public void Stop() to do the cancelling ?

Sotam replied:
I think the latter option seems the most appropriate one 😄

dougdellolio commented on GDAXSharp/GDAXClient.cs:
```diff
@@ -73,5 +78,7 @@ public class GDAXClient
 
         public UserAccountService UserAccountService { get; }
+
+        public WebSocket.WebSocket WebSocket { get; }
```
cool 😄

dougdellolio commented on GDAXSharp/WebSocket/Types/ChannelType.cs:
```diff
@@ -0,0 +1,20 @@
+        [EnumMember(Value = "user")]
+        User
+    }
+}
```
nl

dougdellolio commented on GDAXSharp/WebSocket/Models/Response/Level2.cs:
```diff
@@ -0,0 +1,14 @@
+{
+    public class Level2 : BaseMessage
+    {
+        public string ProductId { get; set; }
```
i think this should be a `ProductType`. just add the `[JsonConverter(typeof(StringEnumConverter))]` above it

dougdellolio commented on GDAXSharp/Network/HttpRequest/HttpRequestMessageService.cs:
```diff
@@ -1,39 +1,32 @@
+        public HttpRequestMessageService(
+            IAuthenticator authenticator,
+            IClock clock,
+            bool sandBox) : base(authenticator, clock, sandBox)
```
hate to be a pain about styling but please new line this:

````
public HttpRequestMessageService(
           IAuthenticator authenticator,
           IClock clock,
           bool sandBox) 
                  : base(authenticator, clock, sandBox)
````

just like to keep the codebase as consistent and clean as possible

dougdellolio commented on GDAXSharp/WebSocket/WebSocket.cs:
```diff
@@ -0,0 +1,157 @@
+            }
+
+            var json = e.Message;
+            var response = JsonConvert.DeserializeObject<BaseMessage>(json);
```
seeing an error being thrown here when you leave the list of `Channels` as null.

````
Newtonsoft.Json.JsonSerializationException occurred
  Message=Error converting value "last_match" to type 'GDAXSharp.WebSocket.Types.ResponseType'. Path 'type'
````

Sotam commented on GDAXSharp/Shared/Utilities/AbstractJson.cs:
```diff
@@ -0,0 +1,28 @@
+﻿using Newtonsoft.Json;
```
@chrisw000  @dougdellolio Not really satisfied with this file. For the `Services` and `WebSocket` we need to do some de- and serialization. So I thought, make it abstract and inherit, so it's easy to use without duplicate code. Perhaps a `Shared` class would be better?

chrisw000 replied:
internal static class, with internal static methods to serialize / deserialize?

Sotam commented on GDAXSharp/WebSocket/Types/ResponseType.cs:
```diff
@@ -15,6 +13,8 @@ public enum ResponseType
         Open,
         Done,
         Match,
+        [EnumMember(Value = "last_match")]
```
@chrisw000 @dougdellolio  Is this the best way for deserializing `last_match` to `LastMatch` ?

chrisw000 replied:
@sotam - yep

Sotam commented:
@dougdellolio @chrisw000 is there a function/extension for VS2017/ReSharper to automatically add newline to file endings? 👼

chrisw000 commented:
In the ReSharper Options, Code Editing, C#, Line Breaks and Wrappings, General -> Line feed at end of file - not sure if it only does that on a brand new file though?

Sotam commented:
@dougdellolio @chrisw000  could you please review the latest commit, about the Models? I'm not quite sure if I can use the `GUID` type or should stick to `string`

chrisw000 commented:
I think they're all guids... you've got a System.IO.File.WriteAllText to your C drive left in there.

dougdellolio commented on GDAXSharp/WebSocket/WebSocket.cs:
```diff
@@ -0,0 +1,212 @@
+            var json = e.Message;
+            var response = JsonConfig.DeserializeObject<BaseMessage>(json);
+
+            System.IO.File.WriteAllText($@"C:\Users\Kevin Wilbrink\Desktop\GDAX\{response.Type.ToString()}", json);
```
dont forget to remove

dougdellolio commented on GDAXSharp/GDAXSharp.csproj:
```diff
@@ -35,29 +35,81 @@
+    <Compile Include="Shared\Utilities\JsonConfig.cs" />
+    <Compile Include="WebSocket\Models\Response\Done.cs" />
+    <Compile Include="WebSocket\Models\Response\Error.cs" />
+    <Compile Include="WebSocket\Models\Response\Hearthbeat.cs" />
```
i think your build is failing is because of these files.

Hearthbeat, Match and open

dougdellolio commented on GDAXSharp/WebSocket/WebSocket.cs:
```diff
@@ -0,0 +1,212 @@
+                throw new GDAXSharpWebSocketException("Websocket can't be stopped")
+                {
+                    WebSocket = WebSocketFeed,
+                    StatusCode = WebSocketFeed?.State ?? WebSocketState.None,
```
you can remove the right side of the `??` because the left side will never be null

chrisw000 commented:
RE separate event handlers & being able to select the channels... I can see an argument for it being this way... I might code up a way to handle every channel possibility - but allow the user or app to change what its subscribed to from run to run?

Each event sends out a different object, which will need to be processed in a different way - so if there were one event handler, it would be simpler to hook up, but then I'd need to write some code to work out which object I'd just got back, and then probably send that object off to a new method for processing... which takes it back to being like having separate event handlers?

So swings and roundabouts....?

When the socket is started it could check to see if there is a delegate attached to the event handler for the particular channel - and throw an exception if not?

Sotam commented:
@dougdellolio @chrisw000 Since the objects are different for each response type, I thought this is a nice way to solve the different types. Also I'm not quite sure if there is a better way. Currently I hook the `OnLevel2UpdateReceived` EventHandler to my `UpdateOrderBook` method and the `OnSnapShotReceived` to my `CreateOrderBook` method. But please let me know your opinions.

chrisw000 commented:
@dougdellolio @Sotam 
I think separate event handlers makes sense. 

I can also see what doug is saying.. you wire the events and then tell it what you want - so it seems like it's stating it twice.  

But, I think as its coded currently is is the cleanest way from the callers point of view.

If it were one event handler, the caller needs to do work when receiving each event, in order to send it onto correct place for processing.

Alternatively, if the web socket just subscribed to everything and only sent on data that had events subscribed, there is potentially a load of network activity going on that is not required.

Finally... if the websocket were to inspect the events, to see which had delegates, and used that to determine which channels should be subscribed to automatically.... then the caller would still need a bunch of logic around which events to hook up based on what it wants at a particular run time requirement.

IE - if the calling code wanted the ability to hook up to differing channels at differing times... its most flexible and clean to hook up all possible events, and then tell the websocket which channels you want this time... when starting it..... which is where the code is now?

dougdellolio commented:
ahhh i understand now.. i think I was just understanding the code incorrectly which is on me.

i also see how having one event handler wouldn't necessarily make sense and would add more work for the user

going to do one more pass at this and then merge this away if you have no concerns. i appreciate all of your time @Sotam and @chrisw000. this is great!

dougdellolio commented:
also @Sotam if there are unimplemented bits that are available to the user we should highlight them somewhere (an issue maybe). and not expose them to the user.

dougdellolio commented on GDAXSharp/WebSocket/Models/Response/Received.cs:
```diff
@@ -0,0 +1,27 @@
+
+        public long Sequence { get; set; }
+
+        public System.DateTimeOffset Time { get; set; }
```
can remove this extra `System.`

dougdellolio commented on README.md:
```diff
@@ -20,6 +20,21 @@ var gdaxClient = new GDAXSharp.GDAXClient(authenticator);
 //use one of the services 
 var allAccounts = await gdaxClient.AccountsService.GetAllAccountsAsync();
+
+//use the websocket feed
```
i'd argue this deserves its own box 😄 one for using the services and one for the websocket feed

dougdellolio commented on README.md:
```diff
@@ -81,6 +96,20 @@ var allAccounts = await gdaxClient.AccountsService.GetAllAccountsAsync();
 ###### User Account ######
 - GetTrailingVolumeAsync() - get 30-day trailing volume for all products
 
+###### Websocket Feed ######
```
not a huge deal but i think this should also have its own section. doesn't really belong with the services

dougdellolio commented on GDAXSharp/Exceptions/GDAXSharpWebSocketException.cs:
```diff
@@ -0,0 +1,26 @@
+        {
+        }
+
+        public GDAXSharpWebSocketException(string message) : base(message)
```
nl this base
````
 public GDAXSharpWebSocketException(string message)
     : base(message)
````

dougdellolio commented on GDAXSharp/Exceptions/GDAXSharpWebSocketException.cs:
```diff
@@ -0,0 +1,26 @@
+        {
+        }
+
+        public GDAXSharpWebSocketException(string message, Exception inner) : base(message, inner)
```
here too

Sotam commented:
Awesome! I'll make the final adjustments in a few hours and do a final commit 😃

Sotam commented:
I've updated the README, please let me know if this is better and currently there is no option to unsubscribe. You've to stop the websocket and (re-)start without the channels you don't want to listen.

Sotam commented:
Fixed a bug, when calling the `WebSocket` but not the `Start` method, a `nullexception` will be thrown when exiting (`WebSocket.Dispose();`) 😄
